### PR TITLE
fixed all bugs in Profile view and added new features

### PR DESCRIPTION
**Pull request by marpuz**
- fixed bugs in Profile view
- added new toasts
- added modal for password confirm

### CODE REVIEW

File: backend/src/controller/user.ts [hunk 1]
```diff
@@ -14,6 +14,34 @@ module.exports.getAllUsers = async (req: Request, res: Response) => {
...
+        const response = await bcrypt.compare(password, userAccount.password);
+        return response;
+    } catch (e: any) {
+        return e;
+    }
+};
+
+module.exports.confirmUser = async (req: Request, res: Response) => {
+    res.setHeader("Access-Control-Allow-Credentials", "true");
+    if (req.user === undefined) return res.status(401).json({ message: "Not authenticated" });
```
Comment (dorotanowak00) [inline backend/src/controller/user.ts]: Suggestion: if (!req.user) {...} will do the same

File: frontend/src/components/Modal/ConfirmModal/ConfirmModal.tsx [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+    const handleSubmit = async (e: React.FormEvent) => {
+        e.preventDefault();
+        await confirmUser(password)
+            .then((data) => {
+                if (data) functionToConfirm();
+            })
+            .catch((err) => {
+                setErrorMessage("Password is incorrect");
+                toast.error("Something gone wrong!", {
```
Comment (dorotanowak00) [inline frontend/src/components/Modal/ConfirmModal/ConfirmModal.tsx]: Suggestion: put the same error message in toast

File: frontend/src/components/Modal/ProfileModal/ProfileModal.tsx [hunk 3]
```diff
@@ -10,13 +10,22 @@ import { getProfilesForLoggedUser } from "store/profile";
 import { inputsArray } from "./InputsArray";
 import { ProfileModalProps } from "interfaces";
 import { AxiosResponse } from "axios";
+import { ProfilePicture } from "components/profilePicture/ProfilePicture";
```
Comment (dorotanowak00) [inline frontend/src/components/Modal/ProfileModal/ProfileModal.tsx]: I believe that path "components" will be enough.

File: frontend/src/components/Modal/ProfileModal/ProfileModal.tsx [hunk 4]
```diff
@@ -39,54 +48,140 @@ export const ProfileModal = ({
...
             hideMessage();
             changeVisiblity();
         },
     });
 
+    const onImgSrcChange = (arg: InputFileChange) => {
+        uploadProfilePhoto(arg.productFromData);
+    };
+
+    const onFileSelect = (file: any) => {
```
Comment (dorotanowak00) [inline frontend/src/components/Modal/ProfileModal/ProfileModal.tsx]: Suggestion: specify the type

File: frontend/src/components/Modal/ProfileModal/ProfileModal.tsx [hunk 5]
```diff
@@ -39,54 +48,140 @@ export const ProfileModal = ({
...
+        const foundItem = fileType.find((item) => item.label === file.type);
+        if (!foundItem) return;
+        if (file.size > MAX_FILE_SIZE) return;
+        const imageForm = new FormData();
+        imageForm.append("file", file);
+        onImgSrcChange({ productImageUrl: URL.createObjectURL(file), productFromData: imageForm });
+    };
+
+    const onFileInputChange = (e: any) => {
```
Comment (dorotanowak00) [inline frontend/src/components/Modal/ProfileModal/ProfileModal.tsx]: Suggestion: specify the type

File: frontend/src/components/Modal/ProfileModal/ProfileModal.tsx [hunk 6]
```diff
@@ -39,54 +48,140 @@ export const ProfileModal = ({
...
+        const fil = e.target.files[0];
+        onFileSelect(fil);
+    };
+
+    const uploadProfilePhoto = async (file: any) => {
```
Comment (dorotanowak00) [inline frontend/src/components/Modal/ProfileModal/ProfileModal.tsx]: As in above comment

File: frontend/src/components/Modal/ProfileModal/ProfileModal.tsx [hunk 7]
```diff
@@ -96,9 +191,25 @@ export const ProfileModal = ({
...
+                        className={styles.deleteButton}
+                        onClick={() => {
+                            changeVisiblity();
+                            openConfirmModal && openConfirmModal();
+                        }}
+                    >
+                        Delete
+                    </button>
+                ) : (
+                    <></>
```
Comment (dorotanowak00) [inline frontend/src/components/Modal/ProfileModal/ProfileModal.tsx]: Suggestion: maybe you can use short circuit instead of returning <></> ?

File: frontend/src/views/Profile/Profile.tsx [hunk 8]
```diff
@@ -14,6 +14,16 @@ import { getProfile } from "API/UserService/profile";
 import { ProfileAuctions } from "./ProfileAuctions/ProfileAuctions";
 import { AuctionItem } from "interfaces";
 import { getUsersAuctions } from "API/UserService/auctions";
+import { UploadCoverPhotoModal } from "components/Modal/UploadCoverPhotoModal/UploadCoverPhotoModal";
+import { ConfirmModal } from "components/Modal/ConfirmModal/ConfirmModal";
```
Comment (dorotanowak00) [inline frontend/src/views/Profile/Profile.tsx]: Suggestion: you can shorten this to: "components/Modal

File: frontend/src/API/UserService/confirmUser.ts [hunk 9]
```diff
@@ -0,0 +1,9 @@
+import axiosInstance from "services/axios";
+
+export const confirmUser = async (password: string) => {
+    const response = await axiosInstance.post(
+        `/user/confirm`,
+        JSON.stringify({ password: password })
+    );
+    return response;
+};
```
Comment (GabrLit) [inline frontend/src/API/UserService/confirmUser.ts]: This may be implicitly returned, i don't know if that would be more readable: 
```tsx
export const confirmUser = async (password: string) =>  await axiosInstance.post(
        `/user/confirm`,
        JSON.stringify({ password: password })
    );
```

File: frontend/src/API/UserService/profile.ts [hunk 10]
```diff
@@ -12,3 +12,14 @@ export const getProfile = async (profileID: string) => {
...
     }
 };
+
+export const deleteProfile = async (profileID: string) => {
+    try {
+        const response: AxiosResponse<ProfileInterface> = await axiosInstance.delete(
+            `/profiles/${profileID}`
+        );
+        return response.data;
+    } catch (error: any) {
```
Comment (GabrLit) [inline frontend/src/API/UserService/profile.ts]: message can be destructured here: 
```tsx
catch ({message})

File: frontend/src/components/Modal/ConfirmModal/ConfirmModal.tsx [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    toastName,
+}: ConfirmModalProps) => {
+    const [password, setPassword] = useState("");
+    const [errorMessage, setErrorMessage] = useState(" ");
+    const navigate = useNavigate();
+    const dispatch = useAppDispatch();
+    const userProfiles = useAppSelector((state) => state.profiles.profiles);
+
+    const handleSubmit = async (e: React.FormEvent) => {
+        e.preventDefault();
```
Comment (GabrLit) [inline frontend/src/components/Modal/ConfirmModal/ConfirmModal.tsx]: you don't really have to pass event if you handle this before this function ie. 
```tsx
onSubmit = {(e)=>{e.preventDefault; handleSubmit()}}
```
this will make handleSubmit fn clearer

File: frontend/src/components/Modal/ConfirmModal/ConfirmModal.tsx [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+export const ConfirmModal = ({
+    changeVisiblity,
+    profileID,
+    userID,
+    functionToConfirm,
+    toastName,
+}: ConfirmModalProps) => {
+    const [password, setPassword] = useState("");
+    const [errorMessage, setErrorMessage] = useState(" ");
```
Comment (GabrLit) [inline frontend/src/components/Modal/ConfirmModal/ConfirmModal.tsx]: Why is there a whitespace

File: frontend/src/components/Modal/ConfirmModal/ConfirmModal.tsx [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    const navigate = useNavigate();
+    const dispatch = useAppDispatch();
+    const userProfiles = useAppSelector((state) => state.profiles.profiles);
```
Comment (GabrLit) [inline frontend/src/components/Modal/ConfirmModal/ConfirmModal.tsx]: You can prepare this selector inside slice, so you dont have to pass state here: 
Slice: 
```tsx
export const selectProfiles = ({ profileSlice }: RootState) => {
  return profileSlice;
};

Invoke: 
const userProfiles = useAppSelector(selectProfiles);

File: frontend/src/components/Modal/ConfirmModal/ConfirmModal.tsx [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+                if (data) functionToConfirm();
+            })
+            .catch((err) => {
+                setErrorMessage("Password is incorrect");
+                toast.error("Something gone wrong!", {
+                    type: "error",
+                    isLoading: false,
+                    autoClose: 2500,
+                    closeOnClick: true,
+                });
```
Comment (GabrLit) [inline frontend/src/components/Modal/ConfirmModal/ConfirmModal.tsx]: You can prepare a wrapper for toasts configs so not to pass values every time

File: frontend/src/components/Modal/ProfileModal/ProfileModal.tsx (hunk 3 above)
Comment (GabrLit) [reply frontend/src/components/Modal/ProfileModal/ProfileModal.tsx]: Yes, provided you have index.ts that exports this function then path can be shortened

File: frontend/src/components/Modal/ProfileModal/ProfileModal.tsx [hunk 15]
```diff
@@ -39,54 +48,140 @@ export const ProfileModal = ({
...
-              twitterUrl: activeProfile?.twitterUrl,
-              facebookUrl: activeProfile?.facebookUrl,
+              profileName: profile?.profileName,
+              profilePicture: profile?.profilePicture,
+              about: profile?.about,
+              websiteUrl: profile?.websiteUrl,
+              instagramUrl: profile?.instagramUrl,
+              twitterUrl: profile?.twitterUrl,
+              facebookUrl: profile?.facebookUrl,
           };
```
Comment (GabrLit) [inline frontend/src/components/Modal/ProfileModal/ProfileModal.tsx]: Having ternary operator here is hard to read

File: frontend/src/components/Modal/ProfileModal/ProfileModal.tsx [hunk 16]
```diff
@@ -39,54 +48,140 @@ export const ProfileModal = ({
...
+                              type: "error",
+                              isLoading: false,
+                              autoClose: 2500,
+                              closeOnClick: true,
+                          });
+                      });
             hideMessage();
             changeVisiblity();
         },
     });
```
Comment (GabrLit) [inline frontend/src/components/Modal/ProfileModal/ProfileModal.tsx]: Ternary operator mess here too, hard to read

Comment (GabrLit) [reply frontend/src/components/Modal/ProfileModal/ProfileModal.tsx]: You can just create two ifs here 
if (isNew)
if (!isNew)

Comment (GabrLit) [reply frontend/src/components/Modal/ProfileModal/ProfileModal.tsx]: Or use if else statement

Comment (GabrLit) [reply frontend/src/components/Modal/ProfileModal/ProfileModal.tsx]: This will flatten the structure, also prepare toast config wrapper, redundant code

File: frontend/src/components/Modal/ProfileModal/ProfileModal.tsx (hunk 4 above)
Comment (GabrLit) [reply frontend/src/components/Modal/ProfileModal/ProfileModal.tsx]: 👍

File: frontend/src/components/Modal/ProfileModal/ProfileModal.tsx [hunk 17]
```diff
@@ -39,54 +48,140 @@ export const ProfileModal = ({
...
+        const foundItem = fileType.find((item) => item.label === file.type);
+        if (!foundItem) return;
+        if (file.size > MAX_FILE_SIZE) return;
+        const imageForm = new FormData();
+        imageForm.append("file", file);
+        onImgSrcChange({ productImageUrl: URL.createObjectURL(file), productFromData: imageForm });
+    };
+
+    const onFileInputChange = (e: any) => {
+        const fil = e.target.files[0];
```
Comment (GabrLit) [inline frontend/src/components/Modal/ProfileModal/ProfileModal.tsx]: not a typo here? what is fil?

File: frontend/src/components/Modal/ProfileModal/ProfileModal.tsx [hunk 18]
```diff
@@ -39,54 +48,140 @@ export const ProfileModal = ({
...
+                closeOnClick: true,
+            })
+        );
+        if (!uploadImage) return null;
+        toast.update(uploadCoverPhotoToast, {
+            render: "Photo uploaded successfully",
+            type: "success",
+            isLoading: false,
+            autoClose: 2500,
+            closeOnClick: true,
```
Comment (GabrLit) [inline frontend/src/components/Modal/ProfileModal/ProfileModal.tsx]: Again, move config to toast wrapper

File: frontend/src/components/Modal/ProfileModal/ProfileModal.tsx (hunk 7 above)
Comment (GabrLit) [reply frontend/src/components/Modal/ProfileModal/ProfileModal.tsx]: 🥇 👍

File: frontend/src/components/Modal/ProfileModal/ProfileModal.tsx [hunk 19]
```diff
@@ -96,9 +191,25 @@ export const ProfileModal = ({
...
+                        Delete
+                    </button>
+                ) : (
+                    <></>
+                )}
+                <button type="submit" className={styles.submitButton}>
+                    {isNew ? <span>Create new profile!</span> : <span>Update profile!</span>}
+                </button>
+            </div>
             {response ? <span className={styles.response}>{response}</span> : null}
```
Comment (GabrLit) [inline frontend/src/components/Modal/ProfileModal/ProfileModal.tsx]: short circut here instead of ternary

File: frontend/src/components/Modal/UploadCoverPhotoModal/UploadCoverPhotoModal.tsx [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+    setProfile,
+}: UploadCoverPhoto) => {
+    const [file, setFile] = useState<FormData>(new FormData());
+    const [formState, setFormState] = useState({
+        imageUrl: profile?.coverPicture,
+        formData: new FormData(),
+    });
+
+    const onFormSubmit = async () => {
+        if (formState.formData === undefined || formState.imageUrl === "")
```
Comment (GabrLit) [inline frontend/src/components/Modal/UploadCoverPhotoModal/UploadCoverPhotoModal.tsx]: if (!formState.formData || !formState.imageUrl)

File: frontend/src/components/Modal/UploadCoverPhotoModal/UploadCoverPhotoModal.tsx [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+            return toast.error("Something is wrong with image...", { autoClose: 2500 });
+        const uploadCoverPhotoToast = toast.loading("Uploading photo...");
+        const uploadImage = await uploadFile(file).catch(() =>
+            toast.update(uploadCoverPhotoToast, {
+                render: "Something is wrong with image!",
+                type: "error",
+                isLoading: false,
+                autoClose: 2500,
+                closeOnClick: true,
```
Comment (GabrLit) [inline frontend/src/components/Modal/UploadCoverPhotoModal/UploadCoverPhotoModal.tsx]: As above, redundant config

File: frontend/src/components/navbarDropDown/NavbarDropDown.tsx [hunk 22]
```diff
@@ -14,6 +14,23 @@ import { Link, useNavigate } from "react-router-dom";
...
+                setActiveDropdownButton(false);
+            }
+        }
+        document.addEventListener("mousedown", handleClickOutside);
+        return () => {
+            document.removeEventListener("mousedown", handleClickOutside);
+        };
+    }, [ref]);
+}
+
```
Comment (GabrLit) [inline frontend/src/components/navbarDropDown/NavbarDropDown.tsx:33]: It is named useOutsideAlerter which indicates it is a hook. So it should be moved to specific folder "hooks"

Comment (GabrLit) [inline frontend/src/components/navbarDropDown/NavbarDropDown.tsx]: Can this be moved to separate folder 'hooks'?

File: frontend/src/components/navbarDropDown/NavbarDropDown.tsx [hunk 23]
```diff
@@ -1,6 +1,6 @@
 import styles from "./NavbarDropDown.module.scss";
 import { useAppSelector } from "store/store";
-import { useState } from "react";
+import { useState, useCallback, useEffect, useRef } from "react";
 import SignOutIcon from "../../assets/SignOutIcon.svg";
```
Comment (GabrLit) [inline frontend/src/components/navbarDropDown/NavbarDropDown.tsx]: relative import?

File: frontend/src/views/Profile/Profile.tsx [hunk 24]
```diff
@@ -14,6 +14,16 @@ import { getProfile } from "API/UserService/profile";
...
+import { deleteProfile } from "../../API/UserService/profile";
```
Comment (GabrLit) [inline frontend/src/views/Profile/Profile.tsx]: relative import?

File: frontend/src/views/Profile/Profile.tsx [hunk 25]
```diff
@@ -31,9 +41,13 @@ export function Profile() {
     const [auctions, setAuctions] = useState<AuctionItem[]>([]);
     const [selectedProfileDisplay, setSelectedProfileDisplay] = useState<string>("onsale");
     const [isModalVisible, setIsModalVisible] = useState(false);
+    const [isUploadCoverPhotoModalVisible, setIsUploadCoverPhotoModalVisible] = useState(false);
+    const [isConfirmModalVisible, setIsConfirmModalVisible] = useState(false);
     const user = useAppSelector((state) => state.user);
     const activeProfile = useAppSelector((state) => state.activeProfile);
     const usersProducts = useAppSelector((state) => state.userProducts.products);
+    const userProfiles = useAppSelector((state) => state.profiles.profiles);
```
Comment (GabrLit) [inline frontend/src/views/Profile/Profile.tsx]: refer to above, can be shortened by exporting selector from slice

File: frontend/src/views/Profile/Profile.tsx [hunk 26]
```diff
@@ -42,14 +56,17 @@ export function Profile() {
             setAuctions(await getUsersAuctions(profileID));
             dispatch(fetchUserProducts(profileID));
         })();
```
Comment (GabrLit) [inline frontend/src/views/Profile/Profile.tsx]: https://dev.to/valeriavg/comment/1f5m2

Do not use raw async functions inside useEffect, declare the function and then invoke

File: frontend/src/views/Profile/Profile.tsx [hunk 27]
```diff
@@ -42,14 +56,17 @@ export function Profile() {
...
     const profileDisplay = (displayOption: string) => {
         switch (displayOption) {
             case "created":
                 return (
                     <CreatedItems
                         usersProducts={usersProducts}
+                        setAuctions={async () =>
+                            profileID && setAuctions(await getUsersAuctions(profileID))
+                        }
                         profileID={profileID ? profileID : ""}
```
Comment (GabrLit) [inline frontend/src/views/Profile/Profile.tsx]: profileID || ""

File: frontend/src/views/Profile/Profile.tsx [hunk 28]
```diff
@@ -58,17 +75,59 @@ export function Profile() {
...
+    };
+
     return (
         <div className={styles.profileContainer}>
             <div className={styles.coverPhotoWrapper}>
-                <img className={styles.coverPhoto} alt="Cover" src={profile?.coverPicture} />
+                <img
+                    className={styles.coverPhoto}
+                    alt="Cover"
+                    src={profile?.coverPicture ? profile.coverPicture : defaultCoverPicture}
```
Comment (GabrLit) [inline frontend/src/views/Profile/Profile.tsx:118]: Short circuit here

**Review by GabrLit — CHANGES_REQUESTED** [review]
Yes, :D some changes

**Review by GabrLit — APPROVED** [review]
LGTM
